File: data/src/main/java/com/somename/data/repository/datasource/mapper/RootEntityMapper.java
package com.somename.data.repository.datasource.mapper;


import com.somename.data.сontent.PostEntity;
import com.somename.data.сontent.RootEntity;
import com.somename.domain.model.Post;
import com.somename.domain.model.Root;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class RootEntityMapper extends Mapper<Root, RootEntity> {

    private final ThumbnailEntityMapper mThumbnailEntityMapper;

    @Inject
    public RootEntityMapper(ThumbnailEntityMapper thumbnailEntityMapper) {
        mThumbnailEntityMapper = thumbnailEntityMapper;
    }

    @Override
    public RootEntity map(Root root) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Root reverseMap(RootEntity rootEntity) {
        Root root = new Root();
        List<Post> posts = new ArrayList<>();
        for (PostEntity postEntity: rootEntity.getPost()) {
            Post post = new Post();
            post.setName(postEntity.getName());
            post.setTagline(postEntity.getTagline());
            post.setThumbnail(mThumbnailEntityMapper.reverseMap(postEntity.getThumbnail()));
            post.setId(postEntity.getId());
            post.setVotesCount(postEntity.getVotesCount());
            posts.add(post);
        }
        root.setPost(posts);
        return root;
    }
}

File: data/src/main/java/com/somename/data/repository/datasource/mapper/RootEntityMapper.kt
package com.somename.data.repository.datasource.mapper


import com.somename.data.сontent.RootEntity
import com.somename.domain.model.Post
import com.somename.domain.model.Root

import java.util.ArrayList

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RootEntityMapper @Inject
constructor(private val mThumbnailEntityMapper: ThumbnailEntityMapper) : Mapper<Root, RootEntity>() {

    override fun map(value: Root): RootEntity {
        throw UnsupportedOperationException()
    }

    override fun reverseMap(value: RootEntity): Root {
        val root = Root()
        val posts = ArrayList<Post>()
        value.post?.forEach { postEntity ->
            val post = Post()
            post.name = postEntity.name
            post.tagline = postEntity.tagline
            post.thumbnail = postEntity.thumbnail?.let { mThumbnailEntityMapper.reverseMap(it) }
            post.id = postEntity.id
            post.votesCount = postEntity.votesCount
            posts.add(post)
        }
        root.post = posts
        return root
    }
}
